### PR TITLE
refactor: refactor lookup zone by denom

### LINKED ISSUE #993: performance: refactor denom->zone lookup
In `x/ics/keeper/msg_server.go`, the method `RequestRedemption` does the following, to determine the zone from the denom requested:

Firstly, this logic should be located elsewhere in the keeper such that it can be reused; and secondly, it would be more performant (marginally) to maintain a map of assets -> zone (similar to how we do for ports -> zone) to lookup up the denom with a single kv lookup instead of repeated iterations (this is only a minor improvement as this code is not hit frequently). 

```
	k.IterateZones(ctx, func(_ int64, thisZone *types.Zone) bool {
		if thisZone.LocalDenom == msg.Value.GetDenom() {
			zone = thisZone
			return true
		}
		return false
	})
```

### PR DESCRIPTION
**Pull request by minhngoc274**
## 1. Summary
Fixes #993

## 2.Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## 5. Checklist

- [ ] Does the Readme need to be updated?

## 7. Future Work (optional)

Handle upgrades V010600UpgradeName, iterate zones and add to denom <-> zone mapping

## Summary by CodeRabbit

- **New Features**
	- Enhanced zone management capabilities for interchain staking, including automatic zone mapping updates.
- **Bug Fixes**
	- Improved the redemption request process by optimizing zone retrieval.
- **Tests**
	- Added tests for local denom zone mapping functionality.
- **Refactor**
	- Streamlined zone deletion and associated record management.

### CODE REVIEW

File: x/interchainstaking/keeper/keeper.go [hunk 1]
```diff
@@ -201,6 +201,50 @@ func (k *Keeper) AllPortConnections(ctx sdk.Context) (pcs []types.PortConnection
...
+	k.IterateZones(ctx, func(_ int64, thisZone *types.Zone) bool {
+		if thisZone.LocalDenom == denom {
+			zone = thisZone
+			k.SetLocalDenomZoneMapping(ctx, thisZone)
+			return true
+		}
+		return false
+	})
+	return zone
+}
```
[the remark this answers is not in the document]
Comment (minhngoc274) [reply x/interchainstaking/keeper/keeper.go:220]: we GetLocalDenomZoneMapping and if zone doesn't exist, we will get then set into the map

File: x/interchainstaking/keeper/keeper.go [hunk 2]
```diff
@@ -201,6 +201,50 @@ func (k *Keeper) AllPortConnections(ctx sdk.Context) (pcs []types.PortConnection
...
+	k.cdc.MustUnmarshal(bz, &zone)
+	return &zone, true
+}
+
+// SetLocalDenomZoneMapping set denom <-> zone mapping.
+func (k *Keeper) SetLocalDenomZoneMapping(ctx sdk.Context, zone *types.Zone) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixLocalDenomZoneMapping)
+	bz := k.cdc.MustMarshal(zone)
+	store.Set([]byte(zone.LocalDenom), bz)
+}
```
[the remark this answers is not in the document]
Comment (minhngoc274) [reply x/interchainstaking/keeper/keeper.go:240]: The funtion that call SetLocalDenomZoneMapping is making sure that zone is not nil. Btw it a good suggestion

**Comment by minhngoc274** [comment]
Handle upgrades V010600UpgradeName, iterate zones and add to denom <-> zone mapping. Will do it when the PR#1184 been merged

File: x/interchainstaking/keeper/keeper.go [hunk 3]
```diff
@@ -201,6 +201,53 @@ func (k *Keeper) AllPortConnections(ctx sdk.Context) (pcs []types.PortConnection
 	return pcs
 }
 
+// GetZoneAndUpdateMappingIfNeeded returns zone by denom.
+func (k *Keeper) GetZoneAndUpdateMappingIfNeeded(ctx sdk.Context, denom string) *types.Zone {
```
Comment (joe-bowman) [inline x/interchainstaking/keeper/keeper.go:205]: Can we call this `GetZoneByLocalDenom` please, as it's a little more intuitive. Otherwise, looks great!

**Review by faddat — APPROVED** [review]
Ah, I like this one very much too!

**Pull request merged**